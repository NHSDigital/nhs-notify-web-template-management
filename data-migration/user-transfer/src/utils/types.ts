export type Parameters = {
  environment: string;
  component: string;
  accessKeyId: string;
  secretAccessKey: string;
  userPoolId: string;
  region: string;
  sessionToken: string;
  flag?: string;
};

export type UserData = {
  username: string;
  /*
   * extracted from the user's group starting with 'client:'
   */
  clientId: string;
  /*
   * extracted from the user's 'sub' attribute
   */
  userId: string;
};

export type Template = {
  id: string;
  owner: string;
};

type ToFrom = {
  from: string;
  to: string;
};

export type TransferStage =
  | 's3:copy'
  | 'ddb:transfer'
  | 's3:delete'
  | 'initial'
  | 'finished';

export type TransferStatus = 'success' | 'failed' | 'migrate';

export type UserTransferPlanItem = {
  templateId: string;
  status: TransferStatus;
  stage: TransferStage;
  reason?: string;
  ddb: {
    owner: ToFrom;
  };
  s3: {
    files: ToFrom[];
  };
};

export type UserTransferPlan = {
  total: number;
  tableName: string;
  bucketName: string;
  migrate: {
    count: number;
    plans: UserTransferPlanItem[];
  };
  orphaned: {
    count: number;
    templateIds: string[];
  };
};

export type UserTransferPlanItemResult = {
  success: boolean;
  stage: TransferStage;
  reasons?: string[];
};
