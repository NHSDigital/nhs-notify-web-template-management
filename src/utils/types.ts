export type FormId =
  | 'choose-a-template-type'
  | 'create-nhs-app-template'
  | 'create-email-template'
  | 'create-letter-template'
  | 'create-text-message-template'
  | 'preview-nhs-app-template'
  | 'submit-template'
  | 'create-nhs-app-template-back';

export type FormErrorState = {
  formErrors: string[];
  fieldErrors: Record<string, string[]>;
};

export enum TemplateType {
  NHS_APP = 'NHS_APP',
  SMS = 'SMS',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
}

export type Session = {
  id: string;
  templateType: TemplateType | 'UNKNOWN';
  nhsAppTemplateName: string;
  nhsAppTemplateMessage: string;
};

export type FormState = {
  validationError?: FormErrorState;
  redirect?: string;
};

export type TemplateFormState = FormState & Session;

export type PageProps = {
  params: {
    sessionId: string;
  };
};

export type TemplateSubmittedPageProps = {
  params: {
    templateId: string;
  };
};

export type PageComponentProps = {
  initialState: TemplateFormState;
};

export type SubmitTemplatePageComponentProps = {
  templateName: string;
  sessionId: string;
};

export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
