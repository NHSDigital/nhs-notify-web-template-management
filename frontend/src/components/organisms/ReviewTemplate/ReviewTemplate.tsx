'use client';

import { ZodErrorSummary } from '@molecules/ZodErrorSummary/ZodErrorSummary';
import { NHSNotifyRadioButtonForm } from '@molecules/NHSNotifyRadioButtonForm/NHSNotifyRadioButtonForm';
import { ReviewTemplateProps } from './review-template.types';

export function ReviewTemplate({
  form,
  ...props
}: Readonly<ReviewTemplateProps>) {
  return (
    <>
      {props.sectionHeading && (
        <div className='notify-confirmation-panel nhsuk-heading-l'>
          {props.sectionHeading}
        </div>
      )}
      <ZodErrorSummary errorHeading={form.errorHeading} state={form.state} />
      {props.PreviewComponent}
      <NHSNotifyRadioButtonForm
        formId={form.formId}
        radiosId={form.radiosId}
        action={form.action}
        state={form.state}
        pageHeading={form.pageHeading}
        options={form.options}
        buttonText={form.buttonText}
        legend={{
          isPgeHeading: false,
          size: 'm',
        }}
        csrfToken={form.csrfToken}
      />
    </>
  );
}
