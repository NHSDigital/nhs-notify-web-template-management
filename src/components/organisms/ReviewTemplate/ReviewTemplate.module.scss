.review__heading {
  word-wrap: anywhere;
}
