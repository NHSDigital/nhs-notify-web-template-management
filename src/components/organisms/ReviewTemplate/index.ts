export * from './ReviewTemplate';
export * from './review-template.types';
