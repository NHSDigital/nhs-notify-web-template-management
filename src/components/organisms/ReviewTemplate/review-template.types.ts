import { PreviewTemplate } from '@molecules/PreviewTemplate/PreviewTemplate';
import { NHSNotifyRadioButtonFormProps } from '@molecules/NHSNotifyRadioButtonForm/NHSNotifyRadioButtonForm';

export type ReviewTemplateProps = {
  sectionHeading: string;
  templateName: string;
  details: {
    heading: string;
    text: { id: string; text: string }[];
  };
  form: {
    errorHeading: string;
  } & NHSNotifyRadioButtonFormProps;
  PreviewComponent: React.ReactElement<typeof PreviewTemplate>;
};
