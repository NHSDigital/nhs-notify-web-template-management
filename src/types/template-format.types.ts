export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
