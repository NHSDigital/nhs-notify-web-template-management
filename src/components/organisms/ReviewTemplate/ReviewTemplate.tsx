'use client';

import styles from './ReviewTemplate.module.scss';

import { Details } from 'nhsuk-react-components';
import { ReviewTemplateProps } from './ReviewTemplate.types';
import { ZodErrorSummary } from '@molecules/ZodErrorSummary/ZodErrorSummary';
import { NHSNotifyRadioButtonForm } from '@molecules/NHSNotifyRadioButtonForm/NHSNotifyRadioButtonForm';

export function ReviewTemplate(props: ReviewTemplateProps) {
  const { form } = props;
  return (
    <>
      <ZodErrorSummary errorHeading={form.errorHeading} state={form.state} />
      <h1
        data-testid='preview-message__heading'
        className={styles.review__heading}
      >
        <span
          data-testid='preview-message__heading-caption'
          className='nhsuk-caption-l'
        >
          {props.sectionHeading}
        </span>
        {props.templateName}
      </h1>
      <Details>
        <Details.Summary data-testid='preview-message-details__heading'>
          {props.details.heading}
        </Details.Summary>
        <Details.Text data-testid='preview-message-details__text'>
          {props.details.text.map((val, idx) => (
            <p key={`details-text-${idx}`}>{val}</p>
          ))}
        </Details.Text>
      </Details>
      {props.PreviewComponent}
      <NHSNotifyRadioButtonForm
        formId={form.formId}
        radiosId={form.radiosId}
        action={form.action}
        state={form.state}
        pageHeading={form.pageHeading}
        options={form.options}
        buttonText={form.buttonText}
        legend={{
          isPgeHeading: false,
          size: 'm',
        }}
      />
    </>
  );
}
