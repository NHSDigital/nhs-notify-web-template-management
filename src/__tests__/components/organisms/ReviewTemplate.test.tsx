import { mockDeep } from 'jest-mock-extended';
import { render, screen } from '@testing-library/react';
import { ReviewTemplate } from '@organisms/ReviewTemplate';
import { EmailTemplate, FormState, NHSAppTemplate } from '@utils/types';

describe('ReviewTemplate component', () => {
  beforeEach(jest.resetAllMocks);

  it('matches snapshot', () => {
    const container = render(
      <ReviewTemplate
        sectionHeading='NHS app message template'
        template={mockDeep<NHSAppTemplate>({
          name: 'Example NHS APP template',
        })}
        form={{
          formId: 'preview-form',
          radiosId: 'preview-example',
          errorHeading: '',
          action: '',
          state: {},
          pageHeading: 'Example heading',
          options: [
            { id: 'option-1', text: 'option 1' },
            { id: 'option-2', text: 'option 2' },
          ],
          buttonText: 'Continue',
        }}
        PreviewComponent={<>Preview</>}
      />
    );

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('matches error snapshot', () => {
    const state: FormState = {
      validationError: {
        formErrors: [],
        fieldErrors: {
          exampleError: ['Example error'],
        },
      },
    };
    const container = render(
      <ReviewTemplate
        sectionHeading='NHS app message template'
        template={mockDeep<NHSAppTemplate>({
          name: 'Example NHS APP template',
        })}
        form={{
          formId: 'preview-form',
          radiosId: 'preview-example',
          errorHeading: '',
          action: '',
          state,
          pageHeading: 'Example heading',
          options: [
            { id: 'option-1', text: 'option 1' },
            { id: 'option-2', text: 'option 2' },
          ],
          buttonText: 'Continue',
        }}
        PreviewComponent={<>Preview</>}
      />
    );

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('renders component correctly', () => {
    render(
      <ReviewTemplate
        sectionHeading='Email template'
        template={mockDeep<EmailTemplate>({
          name: 'Example template',
        })}
        form={{
          formId: 'preview-form',
          radiosId: 'preview-example',
          errorHeading: '',
          action: '',
          state: {},
          pageHeading: 'Example heading',
          options: [
            { id: 'option-1', text: 'option 1' },
            { id: 'option-2', text: 'option 2' },
          ],
          buttonText: 'Continue',
        }}
        PreviewComponent={<>Preview</>}
      />
    );

    expect(screen.getByTestId('preview-message__heading')).toHaveTextContent(
      'Example template'
    );

    expect(
      screen.getByTestId('preview-example-form__legend')
    ).toHaveTextContent('Example heading');

    expect(screen.getByTestId('submit-button')).toHaveTextContent('Continue');
  });
});
