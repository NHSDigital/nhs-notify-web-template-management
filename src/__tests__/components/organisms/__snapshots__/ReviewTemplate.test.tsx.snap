// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewTemplate component matches error snapshot 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="error-summary-title"
    class="nhsuk-error-summary"
    role="alert"
    tabindex="-1"
  >
    <h2
      class="nhsuk-error-summary__title"
      data-testid="error-summary"
      id="error-summary-title"
    />
    <ul
      class="nhsuk-list nhsuk-error-summary__list"
    >
      <li>
        <a
          href="#exampleError"
        >
          Example error
        </a>
      </li>
    </ul>
  </div>
  <h1
    class="review__heading"
    data-testid="preview-message__heading"
  >
    <span
      class="nhsuk-caption-l"
      data-testid="preview-message__heading-caption"
    >
      NHS app message template
    </span>
    Example NHS APP template
  </h1>
  <details
    class="nhsuk-details"
  >
    <summary
      class="nhsuk-details__summary"
      data-testid="preview-message-details__heading"
    >
      <span
        class="nhsuk-details__summary-text"
      >
        Details heading
      </span>
    </summary>
    <div
      class="nhsuk-details__text"
      data-testid="preview-message-details__text"
    >
      <p>
        Details text
      </p>
    </div>
  </details>
  Preview
  <form
    action=""
  >
    <input
      name="form-id"
      readonly=""
      type="hidden"
      value="preview-form"
    />
    <div
      class="nhsuk-form-group"
    >
      <fieldset
        class="nhsuk-fieldset"
      >
        <legend
          class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m"
          data-testid="preview-example-form__legend"
        >
          Example heading
        </legend>
        <div
          class="nhsuk-form-group"
        >
          <div
            class="nhsuk-radios"
            id="preview-example"
          >
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="preview-example-option-1--label"
                class="nhsuk-radios__input"
                data-testid="option-1-radio"
                id="preview-example-option-1"
                name="preview-example"
                type="radio"
                value="option-1"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="preview-example-option-1"
                id="preview-example-option-1--label"
              >
                option 1
              </label>
            </div>
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="preview-example-option-2--label"
                class="nhsuk-radios__input"
                data-testid="option-2-radio"
                id="preview-example-option-2"
                name="preview-example"
                type="radio"
                value="option-2"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="preview-example-option-2"
                id="preview-example-option-2--label"
              >
                option 2
              </label>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <button
      aria-disabled="false"
      class="nhsuk-button"
      data-testid="submit-button"
      id="preview-form-submit-button"
      type="submit"
    >
      Continue
    </button>
  </form>
</DocumentFragment>
`;

exports[`ReviewTemplate component matches snapshot 1`] = `
<DocumentFragment>
  <h1
    class="review__heading"
    data-testid="preview-message__heading"
  >
    <span
      class="nhsuk-caption-l"
      data-testid="preview-message__heading-caption"
    >
      NHS app message template
    </span>
    Example NHS APP template
  </h1>
  <details
    class="nhsuk-details"
  >
    <summary
      class="nhsuk-details__summary"
      data-testid="preview-message-details__heading"
    >
      <span
        class="nhsuk-details__summary-text"
      >
        Details heading
      </span>
    </summary>
    <div
      class="nhsuk-details__text"
      data-testid="preview-message-details__text"
    >
      <p>
        Details text
      </p>
    </div>
  </details>
  Preview
  <form
    action=""
  >
    <input
      name="form-id"
      readonly=""
      type="hidden"
      value="preview-form"
    />
    <div
      class="nhsuk-form-group"
    >
      <fieldset
        class="nhsuk-fieldset"
      >
        <legend
          class="nhsuk-fieldset__legend nhsuk-fieldset__legend--m"
          data-testid="preview-example-form__legend"
        >
          Example heading
        </legend>
        <div
          class="nhsuk-form-group"
        >
          <div
            class="nhsuk-radios"
            id="preview-example"
          >
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="preview-example-option-1--label"
                class="nhsuk-radios__input"
                data-testid="option-1-radio"
                id="preview-example-option-1"
                name="preview-example"
                type="radio"
                value="option-1"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="preview-example-option-1"
                id="preview-example-option-1--label"
              >
                option 1
              </label>
            </div>
            <div
              class="nhsuk-radios__item"
            >
              <input
                aria-labelledby="preview-example-option-2--label"
                class="nhsuk-radios__input"
                data-testid="option-2-radio"
                id="preview-example-option-2"
                name="preview-example"
                type="radio"
                value="option-2"
              />
              <label
                class="nhsuk-label nhsuk-radios__label"
                for="preview-example-option-2"
                id="preview-example-option-2--label"
              >
                option 2
              </label>
            </div>
          </div>
        </div>
      </fieldset>
    </div>
    <button
      aria-disabled="false"
      class="nhsuk-button"
      data-testid="submit-button"
      id="preview-form-submit-button"
      type="submit"
    >
      Continue
    </button>
  </form>
</DocumentFragment>
`;
